src/core/execute.c

```diff
@@ -4022,35 +4026,40 @@ static int compile_suggested_paths(const ExecContext *c, const ExecParameters *p
         return 0;
 }
 
-static int exec_parameters_get_cgroup_path(const ExecParameters *params, char **ret) {
-        bool using_subcgroup;
+static int exec_parameters_get_cgroup_path(Unit *u, const ExecParameters *params, char **ret) {
+        const char *subcgroup = NULL;
         char *p;
+        CGroupContext *c;
 
         assert(params);
         assert(ret);
 
         if (!params->cgroup_path)
                 return -EINVAL;
 
+        if (!(c = unit_get_cgroup_context(u)))
+                return -EOPNOTSUPP;
+
         /* If we are called for a unit where cgroup delegation is on, and the payload created its own populated
-         * subcgroup (which we expect it to do, after all it asked for delegation), then we cannot place the control
-         * processes started after the main unit's process in the unit's main cgroup because it is now an inner one,
-         * and inner cgroups may not contain processes. Hence, if delegation is on, and this is a control process,
-         * let's use ".control" as subcgroup instead. Note that we do so only for ExecStartPost=, ExecReload=,
-         * ExecStop=, ExecStopPost=, i.e. for the commands where the main process is already forked. For ExecStartPre=
-         * this is not necessary, the cgroup is still empty. We distinguish these cases with the EXEC_CONTROL_CGROUP
-         * flag, which is only passed for the former statements, not for the latter. */
-
-        using_subcgroup = FLAGS_SET(params->flags, EXEC_CONTROL_CGROUP|EXEC_CGROUP_DELEGATE|EXEC_IS_CONTROL);
-        if (using_subcgroup)
-                p = path_join(params->cgroup_path, ".control");
-        else
-                p = strdup(params->cgroup_path);
+         * subcgroup, then we cannot place the control processes started after the main unit's process in the
+         * unit's main cgroup because it is now an inner one, and inner cgroups may not contain processes.
+         * Hence, if delegation is on, use a subcgroup per configuration.
+         * Note that we do so only for ExecStartPost=, ExecReload=, ExecStop=, ExecStopPost=, i.e. for the
+         * commands where the main process is already forked. For ExecStartPre= this is not necessary, the
+         * cgroup is still empty. We distinguish these cases with the EXEC_CONTROL_CGROUP flag, which is only
+         * passed for the former statements, not for the latter. */
+
+        if (FLAGS_SET(params->flags, EXEC_CONTROL_CGROUP|EXEC_CGROUP_DELEGATE|EXEC_IS_CONTROL))
+                subcgroup = c->delegate_subcgroup ?: DELEGATE_CGROUP_CONTROL;
+        else if (FLAGS_SET(params->flags, EXEC_CONTROL_CGROUP|EXEC_CGROUP_DELEGATE))
+                subcgroup = c->delegate_subcgroup ?: DELEGATE_CGROUP_PAYLOAD;
+
+        p = path_join(params->cgroup_path, subcgroup);
```

Comment:
this is really hard to follow, because subcgroup can now be valid (which kinda is what most people would assume), or is NULL or is "". 

We strive to write readable code, obvious code, not super optimized code... this is not performance sensitive after all.